Concurrent writing into `Saved` causes partial read of XMIR file

`Saved` truncates its target and streams into it, so a file that two `Threaded` workers write at the same time can be read back half-written by a third. `ConcurrentCache` (#5720) doesn't cover this: it serializes the `Cache.apply` call, but every caller reads the target file *after* the guard has been released.

The three call sites all have this shape — guarded write, unguarded read:

- `Parsing.parsed`: `Parsing.tojoVersion(target, refs)` at [`Parsing.java:213`](https://github.com/objectionary/eo/blob/master/eo-maven-plugin/src/main/java/org/eolang/maven/Parsing.java#L213), which does `new XMLDocument(target)` when `refs` is empty (i.e. on a cache hit), and `new Xnav(target)` at [`Parsing.java:214`](https://github.com/objectionary/eo/blob/master/eo-maven-plugin/src/main/java/org/eolang/maven/Parsing.java#L214), which runs on every path
- `Transpiling.transpiled` → `javaGenerated` → `new Xnav(target).element("object")` at [`Transpiling.java:416`](https://github.com/objectionary/eo/blob/master/eo-maven-plugin/src/main/java/org/eolang/maven/Transpiling.java#L416)
- `Linting.lintOne`: `new Xnav(target)` at [`Linting.java:314`](https://github.com/objectionary/eo/blob/master/eo-maven-plugin/src/main/java/org/eolang/maven/Linting.java#L314)

The write itself is [`Saved.java:85-92`](https://github.com/objectionary/eo/blob/master/eo-maven-plugin/src/main/java/org/eolang/maven/Saved.java#L85-L92) — a `TeeInput` into `new OutputTo(target)`, which opens the file truncating and streams the bytes in. There is a window, proportional to the size of the XMIR, in which the file on disk is a valid path to an incomplete document.

### How it shows up

```
java.lang.IllegalArgumentException: Can't parse by com.sun.org.apache.xerces.internal.jaxp.DocumentBuilderImpl,
    most probably the XML is invalid
  at com.jcabi.xml.XMLDocument.<init>(XMLDocument.java:178)
  at org.eolang.maven.Parsing.tojoVersion(Parsing.java:291)
  at org.eolang.maven.Parsing.parsed(Parsing.java:225)
Caused by: org.xml.sax.SAXParseException; systemId: file:.../target/1-parse/main.xmir;
    lineNumber: 1; columnNumber: 1; Premature end of file.
```

I hit this while writing the regression test for #5720 in #5766 — eight `Threaded` workers of one `Parsing` on the same tail path reproduced it on most runs. The message is misleading: nothing is wrong with the XMIR, it was simply read mid-write.

### Precondition

Two workers have to be writing the same target path, which is the same precondition as #5720 and not something a routine build hits — within one mojo the tojos are keyed by identifier, so their targets differ. It becomes reachable when two tojos map to one `Place`. Note that `Transpiling` and `Linting` derive the target from the object name declared *inside* the XMIR (`OnDetailed(OnDefault(...))`) rather than from the tojo identifier, so two different source files declaring the same object name land on the same target — a user error that lints do catch, but which would surface here as `Premature end of file` instead of a lint message.

So: latent rather than urgent, but the failure mode is opaque enough to be worth closing.

### Possible fixes

1. **Write atomically in `Saved`** — stream into a sibling temp file and `Files.move(..., ATOMIC_MOVE)`. Fixes every reader at once, including the ones not listed above, and needs no changes at the call sites. A reader then sees either the old file or the new one, never a truncated one. Costs one extra file creation per save.
2. **Widen the guarded region** so the read of `target` happens under the same lock as the write. Narrower, but it means each of the three classes has to hold the guard across a larger body, and it does nothing for any future reader of these files.

(1) looks clearly better to me — the invariant belongs to whoever writes the file, not to each caller who reads it back. Happy to send a PR if there's agreement on the approach.

### Related

- #5720 — the cache guard this sits next to; fixed in #5766, which does not address this
- #5766 — its `ParsingTest` deliberately catches and counts this exception, so it can assert on the parse count without being flaky. That `catch` can go away once this is fixed.

--- a/eo-maven-plugin/src/main/java/org/eolang/maven/Saved.java
+++ b/eo-maven-plugin/src/main/java/org/eolang/maven/Saved.java
@@ -7,7 +7,9 @@ package org.eolang.maven;
 import com.jcabi.log.Logger;
 import java.io.IOException;
 import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
 import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
 import org.cactoos.Input;
 import org.cactoos.Scalar;
 import org.cactoos.Text;
@@ -20,6 +22,14 @@ import org.cactoos.scalar.LengthOf;
 /**
  * Content saved to the file.
  * Returns path to the file
+ *
+ * <p>The content is streamed into a sibling temporary file first, then moved
+ * onto {@link #target} with {@link StandardCopyOption#ATOMIC_MOVE}. Streaming
+ * straight into {@link #target} would let a concurrent reader (or another
+ * writer racing on the same path) observe it truncated, mid-write (#5873):
+ * an atomic rename means a reader always sees either the previous complete
+ * file or the new one, never a partial one.</p>
+ *
  * @since 0.41.0
  */
 final class Saved implements Scalar<Path> {
@@ -82,14 +92,27 @@ final class Saved implements Scalar<Path> {
                     this.target.getParent()
                 );
             }
-            bytes = new IoChecked<>(
-                new LengthOf(
-                    new TeeInput(
-                        this.content,
-                        new OutputTo(this.target)
+            final Path tmp = Files.createTempFile(
+                this.target.getParent(),
+                this.target.getFileName().toString(),
+                ".tmp"
+            );
+            try {
+                bytes = new IoChecked<>(
+                    new LengthOf(
+                        new TeeInput(
+                            this.content,
+                            new OutputTo(tmp)
+                        )
                     )
-                )
-            ).value();
+                ).value();
+                Files.move(
+                    tmp, this.target,
+                    StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING
+                );
+            } finally {
+                Files.deleteIfExists(tmp);
+            }
             Logger.debug(
                 this, "File %s saved (%d bytes)",
                 this.target, bytes

--- a/eo-maven-plugin/src/main/java/org/eolang/maven/Saved.java
+++ b/eo-maven-plugin/src/main/java/org/eolang/maven/Saved.java
@@ -7,6 +7,7 @@ package org.eolang.maven;
 import com.jcabi.log.Logger;
 import java.io.IOException;
 import java.nio.charset.StandardCharsets;
+import java.nio.file.AtomicMoveNotSupportedException;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.StandardCopyOption;
@@ -106,10 +107,7 @@ final class Saved implements Scalar<Path> {
                         )
                     )
                 ).value();
-                Files.move(
-                    tmp, this.target,
-                    StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING
-                );
+                Saved.moved(tmp, this.target);
             } finally {
                 Files.deleteIfExists(tmp);
             }
@@ -127,5 +125,24 @@ final class Saved implements Scalar<Path> {
             );
         }
         return this.target;
+    }
+
+    /**
+     * Move a temp file onto its target, atomically where the filesystem
+     * supports it, falling back to a plain (still complete-file, just not
+     * guaranteed atomic) move on filesystems that don't.
+     * @param tmp Temp file to move
+     * @param target Destination path
+     * @throws IOException If the move fails
+     */
+    private static void moved(final Path tmp, final Path target) throws IOException {
+        try {
+            Files.move(
+                tmp, target,
+                StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING
+            );
+        } catch (final AtomicMoveNotSupportedException ignored) {
+            Files.move(tmp, target, StandardCopyOption.REPLACE_EXISTING);
+        }
     }
 }

--- a/eo-maven-plugin/src/test/java/org/eolang/maven/SavedTest.java
+++ b/eo-maven-plugin/src/test/java/org/eolang/maven/SavedTest.java
@@ -38,7 +38,7 @@ final class SavedTest {
     }
 
     @Test
-    void neverExposesAPartiallyWrittenFileToAConcurrentReader(
+    void exposesNoPartiallyWrittenFileToConcurrentReader(
         @Mktmp final Path temp
     ) throws Exception {
         final Path target = temp.resolve("shared.txt");

--- a/eo-maven-plugin/src/test/java/org/eolang/maven/SavedTest.java
+++ b/eo-maven-plugin/src/test/java/org/eolang/maven/SavedTest.java
@@ -1,0 +1,67 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2016-2026 Objectionary.com
+ * SPDX-License-Identifier: MIT
+ */
+package org.eolang.maven;
+
+import com.yegor256.Mktmp;
+import com.yegor256.MktmpResolver;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.hamcrest.MatcherAssert;
+import org.hamcrest.Matchers;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+/**
+ * Test case for {@link Saved}.
+ * @since 0.74
+ */
+@ExtendWith(MktmpResolver.class)
+final class SavedTest {
+
+    @Test
+    void savesContentToFile(@Mktmp final Path temp) throws IOException {
+        final Path target = temp.resolve("out.txt");
+        new Saved("hello", target).value();
+        MatcherAssert.assertThat(
+            "the saved file must contain exactly what was written",
+            Files.readString(target, StandardCharsets.UTF_8),
+            Matchers.equalTo("hello")
+        );
+    }
+
+    @Test
+    void neverExposesAPartiallyWrittenFileToAConcurrentReader(
+        @Mktmp final Path temp
+    ) throws Exception {
+        final Path target = temp.resolve("shared.txt");
+        final List<String> variants = IntStream.range(0, 8)
+            .mapToObj(idx -> String.valueOf((char) ('a' + idx)).repeat(500_000))
+            .collect(Collectors.toList());
+        new Saved(variants.get(0), target).value();
+        final List<String> broken = new CopyOnWriteArrayList<>();
+        new Threaded<>(
+            variants,
+            variant -> {
+                new Saved(variant, target).value();
+                final String read = Files.readString(target, StandardCharsets.UTF_8);
+                if (!variants.contains(read)) {
+                    broken.add(read.substring(0, Math.min(read.length(), 40)));
+                }
+                return 1;
+            }
+        ).total();
+        MatcherAssert.assertThat(
+            "a reader racing with concurrent writers must always see a complete variant, never a mix of two",
+            broken,
+            Matchers.empty()
+        );
+    }
+}

